service levels: update connections parameters automatically

Jadw1 opened the pull request:
This patch makes all cql connections update theirs service level parameters automatically when:
- any service level is created or changed
- one role is granted to another
- any service level is attached to/detached from a role

First of all, the patch defines what a service level and an effective service level are https://github.com/scylladb/scylladb/commit/938aa10509fb499928eb0a35e4acc39abcd138b6. No new type of service levels are introduced, the commit only clarifies definitions and names what an effective service level is.
(Effective service level is created by merging all service levels which are attached to all roles granted to the user. It represents exact  values of connection's parameters.)

Previously, to find an effective service level of a user, it required O(n) internal queries: O(n) queries to recursively find all granted roles (`standard_role_manager::query_granted()`) and a query for each role to get its service level (`standard_role_manager::get_attribute()`, which sums to O(n) queries).

Because we want to reload SL parameters for all opened cql connections, we don't want to do O(n) queries for every connection, every time we create or change any service level/grant one role to another/attach or detach a service level to/from a role.

To speed it up, the patch adds another layer of service level controller cache, which stored `role_name -> effective_service_level` mapping. This way finding a effective service level for a role is only a lookup to a map.
Building the new cache requires only 2 queries: one to obtain all role hierarchy one to get all roles' service level.

Fixes scylladb/scylladb#12923

Jadw1 commented on service/raft/group0_state_machine.cc:
```diff
@@ -178,7 +178,15 @@ future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merg
     },
     [&] (write_mutations& muts) -> future<> {
-        return write_mutations_to_database(_sp, cmd.creator_addr, std::move(muts.mutations));
+        bool update_service_levels = std::any_of(muts.mutations.begin(), muts.mutations.end(), [] (auto& mut) {
```
@kbr- @gleb-cloudius
What is your opinion on this?
I need to reload SL configuration when there are changes to `system.role_members` or `system.role_attributes` tables.
We've been discussing it with @piodul and there is an ongoing discussion about raft changes and updating corresponding subsystems, so we've decided to do it this way and ask for your opinion.
I can create another command type for auth tables, but auth has 4 tables and only 2 of them interest me (so scanning of table ids is unavoidable if we don't want to reload SL configuration on any change to auth).

gleb-cloudius replied:
The direction here is the one I want to take for all of the commands instead of having individual c++ types for each one of them.

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -220,6 +220,72 @@ future<> service_level_controller::update_service_levels_cache() {
     });
 }
 
+static std::unordered_set<sstring> get_all_roles(const std::unordered_multimap<sstring, sstring>& hierarchy, const sstring& base_role) {
```
`get_all_roles` doesn't sound like a good name for this function. It does not return all roles that exist. Maybe something more descriptive?

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -220,6 +220,72 @@ future<> service_level_controller::update_service_levels_cache() {
+            ::parallel_for_each(attributes.begin(), attributes.end(), [&hierarchy, &attributes, &effective_sl_map, this] (auto& attribute_entry) -> future<> {
+                auto& base_role = attribute_entry.first;
+
+                auto all_roles = get_all_roles(hierarchy, base_role);
```
It looks like you are going over all roles, for each role you compute all their "descendants" (not sure if that's the proper term here), then you combine them to calculate the effective service level. Is this necessary to always enumerate all descendant and combine?

I know that it's more efficient now that we have everything in memory, but it looks like we could benefit from some memoization: it looks to me that when enumerating service levels, if you encounter a service level for which you already computed the effective service level, you could take that effective service level and not have to push the whole subtree.

IIUC combining service levels is an associative and commutative operation, if so then this should work and should give you O(N) running time for refreshing all service levels.

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -317,12 +317,18 @@ void service_level_controller::stop_update_from_distributed_data() {
 
-future<std::optional<service_level_options>> service_level_controller::find_effective_service_level(auth::role_set roles, include_effective_names include_names) {
+future<std::optional<service_level_options>> service_level_controller::find_effective_service_level(const sstring& role_name) {
+    if (true) { //TODO: fix it, properly determine if service level controller already switched to raft updates
```
It looks like we need the same condition as in your other PR (refreshing not every 10s but on every raft update) - check the cluster feature and the topology coordinator upgrade status.

Jadw1 replied:
It's already done https://github.com/scylladb/scylladb/pull/18758/commits/976d97bc710afddd1abec2c7f23fa5b325e9f19e
I think `!group0_client.in_recovery() && feature_service.raft_service_levels_change && storage_service.get_topology_upgrade_state() == service::topology::upgrade_state_type::done` condition is enough in this place as well

piodul commented on generic_server.cc:
```diff
@@ -40,13 +40,12 @@ connection::~connection()
 }
 
-future<> server::for_each_gently(noncopyable_function<void(connection&)> fn) {
+future<> server::for_each_gently(noncopyable_function<future<>(connection&)> fn) {
```
> generic_server: use async function in for_each_gently()

Missing motivation from the commit description

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -333,9 +339,7 @@ future<std::optional<service_level_options>> service_level_controller::find_effe
-                if (include_names == include_effective_names::yes) {
-                    sl_it->second.slo.init_effective_names(*sl_name);
-                }
+                sl_it->second.slo.init_effective_names(*sl_name);
```
What's the reason for this change? I don't understand why we were doing it in the first place, but I also don't understand why it's OK to remove the `include_names` parameter and set the effective SL names unconditionally.

Jadw1 replied:
Effective names were added later to support `LIST EFFECTIVE SERVICE LEVEL OF <role>`. It wasn't needed anywhere else, so I made it optional to include the names.
But in effective service levels cache the effective names are always included, so also made it unconditional in old path (without effective sl cache) in `find_effective_service_level()`

piodul commented on transport/controller.cc:
```diff
@@ -332,14 +332,22 @@ future<> controller::subscribe_server(sharded<cql_server>& server) {
     return server.invoke_on_all([this] (cql_server& server) {
         _mnotifier.local().register_listener(server.get_migration_listener());
         _lifecycle_notifier.local().register_subscriber(server.get_lifecycle_listener());
+        if (!_used_by_maintenance_socket) {
```
> Make cql server (but not maintenance server) is subscribed to qos
> configuration change.

Why is it necessary not to subscribe to the events in maintenance mode? I feel this should be explained in the commit message.

piodul commented on transport/controller.cc:
```diff
@@ -332,14 +332,22 @@ future<> controller::subscribe_server(sharded<cql_server>& server) {
 }
 
 future<> controller::unsubscribe_server(sharded<cql_server>& server) {
     return server.invoke_on_all([this] (cql_server& server) {
         return _mnotifier.local().unregister_listener(server.get_migration_listener()).then([this, &server]{
-            return _lifecycle_notifier.local().unregister_subscriber(server.get_lifecycle_listener());
+            return _lifecycle_notifier.local().unregister_subscriber(server.get_lifecycle_listener()).then([this, &server] {
+                if (!_used_by_maintenance_socket) {
+                    return _sl_controller.local().unregister_subscriber(server.get_qos_configuration_listener());
+                }
+                return make_ready_future<>();
+            });
         });
     });
```
Looks like a good opportunity to coroutinize this function (in a separate commit that would precede this commit)

piodul commented:
@Jadw1 merging this PR will fix https://github.com/scylladb/scylladb/issues/12923, right?

Jadw1 commented:
> @Jadw1 merging this PR will fix #12923, right?
Yes

piodul commented on auth/role_manager.hh:
```diff
@@ -76,6 +76,7 @@ public:
 };
 
 using role_set = std::unordered_set<sstring>;
+using direct_roles_map = std::multimap<sstring, sstring>;
```
I don't think this name is descriptive enough. "Direct roles map" does not tell much about the relationship between keys and value.

How about `role_to_direct_grantees_map`? I don't like it too much either because it requires non-zero concentration in order to parse it, but I think it does describe the relationship correctly.

Jadw1 replied:
`grantees` isn't a good name because (role1, role2) entry in the map means permissions of role2 are subset of role1's premissions.
Changed it to `role_to_direct_granted_map`.

piodul commented on service/qos/service_level_controller.hh:
```diff
@@ -118,6 +118,7 @@ private:
     static constexpr shard_id global_controller = 0;
 
     std::map<sstring, service_level> _service_levels_db;
+    std::map<sstring, service_level_options> _effective_service_levels_db;
```
The term "effective service level" is not too clear without reading an explanation, and here it's not too clear what the key is supposed to be. It's particularly confusing that `_effective_service_levels_db` looks very similar (the same key type, defined right next to it) to `_service_levels_db`, but the first one is indexed by _role_ name and the other is indexed by _service level_ name.

I'd appreciate a short comment around this variable explaining what the key is.

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -220,6 +223,64 @@ future<> service_level_controller::update_service_levels_cache() {
+            auto [it, it_end] = hierarchy.equal_range(role);
+            while (it != it_end) {
+                auto& subrole = it->second;
+                if (effective_sl_map.contains(subrole)) {
+                    auto& sub_sl = effective_sl_map.at(subrole);
```
Nit: why not do `if (auto effective_sl_it = effective_sl_map.find(subrole); effective_sl_it != effective_sl_map.end()) {}` and avoid double lookup?

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -9,6 +9,7 @@
 #include <boost/algorithm/string/join.hpp>
 #include <chrono>
 
+#include <ranges>
```
Is this header necessary?

piodul commented on service/raft/group0_state_machine.cc:
```diff
@@ -141,20 +141,27 @@ static future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_addr
 }
 
 group0_state_machine::modules_to_reload group0_state_machine::get_modules_to_reload(const std::vector<canonical_mutation>& mutations) {
+    using sys = db::system_keyspace;
```
No, please do not introduce unfamiliar shortcuts like this. I don't think this improves readability by much.

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -281,6 +281,14 @@ future<> service_level_controller::update_effective_service_levels_cache() {
 
+future<> service_level_controller::update_cache(update_only_effective_cache update_only_effective_cache) {
+    assert(this_shard_id() == global_controller);
+    if (!update_only_effective_cache) {
+        co_await update_service_levels_cache();
+    }
+    co_await update_effective_service_levels_cache();
```
These will take the `notifications_serializer` semaphore twice. Can something break if a rogue `add_service_level` executes between them?

Jadw1 replied:
No, it's not a problem.
`add_service_level` adds the entry to service level cache, so it's up-to-date.

piodul commented on service/qos/service_level_controller.hh:
```diff
@@ -45,6 +46,8 @@ struct service_level {
      bool is_static;
 };
 
+using update_only_effective_cache = bool_class<class update_only_effective_cache_tag>;
```
I think it would be better to introduce a type called `update_service_level_cache` and have it set to `yes` by default as the argument to `update_cache`.

Parameters such as `update_only_effective_cache` have less chance to be orthogonal with future parameters due to the "only" part; imagine if you want to add more flags like this that will control reloading of other stuff.

piodul commented on service/raft/group0_state_machine.cc:
```diff
@@ -141,20 +141,27 @@ static future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_addr
-        if (mut.column_family_id() == db::system_keyspace::service_levels_v2()->id()) {
+        auto id = mut.column_family_id();
+
+        if (id == sys::service_levels_v2()->id()) {
             modules.service_levels_cache = true;
```
If the service levels table is modified then we need to update the cache, right? I think that `modules.service_levels_effective_cache` should be set to `true` in that case.

piodul commented on service/raft/group0_state_machine.cc:
```diff
@@ -141,20 +141,27 @@ static future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_addr
 future<> group0_state_machine::reload_modules(modules_to_reload modules) {
-    if (modules.service_levels_cache) {
-        co_await _ss.update_service_levels_cache();
+    if (modules.service_levels_cache) { // this also updates SL effective cache
+        co_await _ss.update_service_levels_cache(qos::update_only_effective_cache::no);
+    } else if (modules.service_levels_effective_cache) {
+        co_await _ss.update_service_levels_cache(qos::update_only_effective_cache::yes);
     }
```
How about:

```c++
if (modules.service_levels_cache || modules.service_levels_effective_cache) {
    co_await _ss.update_service_levels_cache(qos::update_service_level_cache(modules.service_levels_cache));
}
```

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -278,6 +278,7 @@ future<> service_level_controller::update_effective_service_levels_cache() {
         }
 
         _effective_service_levels_db = std::move(effective_sl_map);
+        notify_effective_service_levels_cache_reloaded().get();
```
This code doesn't run in `seastar::async`.

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -305,12 +305,18 @@ void service_level_controller::stop_legacy_update_from_distributed_data() {
-future<std::optional<service_level_options>> service_level_controller::find_effective_service_level(auth::role_set roles, include_effective_names include_names) {
+future<std::optional<service_level_options>> service_level_controller::find_effective_service_level(const sstring& role_name) {
+    if (_sl_data_accessor->is_topology_state_upgraded()) {
+        co_return _effective_service_levels_db.contains(role_name) 
+            ? std::optional<service_level_options>(_effective_service_levels_db.at(role_name)) 
+            : std::nullopt;
```
Nit: avoid double lookup by using iterators instead of the `contains` + `at` combo

piodul commented on service/qos/service_level_controller.cc:
```diff
@@ -384,6 +385,14 @@ future<> service_level_controller::notify_service_level_removed(sstring name) {
     co_return;
 }
 
+future<> service_level_controller::notify_effective_service_levels_cache_reloaded() {
+    return seastar::async([this] {
+        _subscribers.thread_for_each([] (qos_configuration_change_subscriber* subscriber) {
+            subscriber->on_effective_service_levels_cache_reloaded().get();
+        });
+    });
+}
+
```
I'm not sure what we gain here by using `seastar::async` and `thread_for_each`. The `on_effective_service_levels_cache_reloaded` method that is called in the loop returns a future, so it is not assume presence of `seastar::async` and does not take advantage of it. `seastar::async` allocates a 128kb stack, so using it is a bit wasteful here.

I suggest removing `seastar::async` and changing this to `_subscriber.for_each`.

piodul reviewed (changes requested):
There are some typos reported by the spellchecker, please fix them.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10598/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10598/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/service_level_controller_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10598/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10807/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10598/)
#### Failed Tests (216/3046):
- [test_auth_no_quorum.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.debug.1)
- [test_auth_raft_snapshot_transfer.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer.debug.1)
- [test_auth_no_quorum.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.debug.2)
- [test_auth_raft_snapshot_transfer.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer.debug.2)
- [test_auth_no_quorum.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_no_quorum_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_no_quorum.debug.3)
- [test_auth_raft_snapshot_transfer.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_no_quorum/Tests___Unit_Tests___test_auth_raft_snapshot_transfer_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_snapshot_transfer.debug.3)
- [test_auth_raft_command_split.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_raft_command_split/Tests___Unit_Tests___test_auth_raft_command_split_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_command_split.debug.1)
- [test_auth_raft_command_split.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_raft_command_split/Tests___Unit_Tests___test_auth_raft_command_split_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_command_split.debug.2)
- [test_auth_raft_command_split.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_raft_command_split/Tests___Unit_Tests___test_auth_raft_command_split_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_raft_command_split.debug.3)
- [test_auth_v2_migration.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10598/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.debug.1)
#### Build Details:
- Duration: 3 hr 32 min
- Builder: i-0c90a97ab9cce3512 (m5ad.8xlarge)

Jadw1 commented on service/qos/service_level_controller.cc:
```diff
@@ -220,6 +223,64 @@ future<> service_level_controller::update_service_levels_cache() {
+
+    return with_semaphore(_global_controller_db->notifications_serializer, 1, [this] () -> future<> {
+        auto& role_manager = _auth_service.local().underlying_role_manager();
+        const auto all_roles = co_await role_manager.query_all();
+        const auto hierarchy = co_await role_manager.query_all_directly_granted();
+        // includes only roles with attached service level
+        const auto attributes = role_manager.query_attribute_for_all("service_level").get();
```
Yeah I know. I didn't notice this in the first place when I've removed seastar::async

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10629/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10629/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/service_level_controller_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10629/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10831/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10629/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2814/)
#### Failed Tests (9/31530):
- [test_auth_v2_migration.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10629/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.debug.1)
- [test_auth_v2_migration.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10629/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.debug.2)
- [test_auth_v2_migration.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10629/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.debug.3)
- [test_auth_v2_migration.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10629/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.dev.1)
- [test_auth_v2_migration.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10629/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.dev.2)
- [test_auth_v2_migration.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10629/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.dev.3)
- [test_auth_v2_migration.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10629/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.release.1)
- [test_auth_v2_migration.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10629/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.release.2)
- [test_auth_v2_migration.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10629/testReport/junit/%28root%29/test_auth_v2_migration/Tests___Unit_Tests___test_auth_v2_migration_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_auth_v2_migration.release.3)
#### Build Details:
- Duration: 5 hr 19 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10695/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10695/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: boost/service_level_controller_test
#### Failed Tests (20/181):
- [test_connections_parameters_auto_update.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.1)
- [test_connections_parameters_auto_update.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.10)
- [test_connections_parameters_auto_update.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.2)
- [test_connections_parameters_auto_update.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.3)
- [test_connections_parameters_auto_update.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.4)
- [test_connections_parameters_auto_update.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.5)
- [test_connections_parameters_auto_update.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.6)
- [test_connections_parameters_auto_update.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.7)
- [test_connections_parameters_auto_update.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.8)
- [test_connections_parameters_auto_update.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10695/testReport/junit/%28root%29/test_raft_service_levels/test_connections_parameters_auto_update_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_connections_parameters_auto_update.debug.9)
#### Build Details:
- Duration: 1 hr 10 min
- Builder: spider2.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10703/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10703/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: boost/service_level_controller_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10703/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2868/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10890/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10703/)
#### Build Details:
- Duration: 5 hr 6 min
- Builder: spider2.cloudius-systems.com

Jadw1 commented:
`iwyu / Analyze #includes in source files (pull_request) ` failed but the error seems unrelated to the PR code
```
  swagger_targets=''
  for f in api/api-doc/*.json; do
    if test "${f#*.}" = json; then
      name=$(basename "$f" .json)
      if test $name != swagger20_header; then
        swagger_targets+=" scylla_swagger_gen_$name"
      fi
    fi
  done
  cmake                                         \
    --build build                               \
     --target seastar_http_request_parser       \
     --target idl-sources                       \
     --target $swagger_targets
  shell: sh -e {0}
  env:
    BUILD_TYPE: RelWithDebInfo
    BUILD_DIR: build
    CLEANER_OUTPUT_PATH: build/clang-include-cleaner.log
    CLEANER_DIRS: test/unit exceptions alternator api auth cdc compaction
ninja: error: unknown target 'scylla_swagger_gen_cql_server_test'
Error: Process completed with exit code 1.
```
@yaronkaikov

Jadw1 commented:
Changelog:
- fixed test (in some raft auth tests: some roles have assigned service levels but the service levels weren't actually created; moved service levels migration to raft before auth migration because the new effective service levels cache depends on the 1st level SL cache and on auth)
- added cql server testing API
- added new test to validate connections' parameters are automatically updated
- addressed latest comments

yaronkaikov commented:
> `iwyu / Analyze #includes in source files (pull_request) ` failed but the error seems unrelated to the PR code
> 
> ```
>   swagger_targets=''
>   for f in api/api-doc/*.json; do
>     if test "${f#*.}" = json; then
>       name=$(basename "$f" .json)
>       if test $name != swagger20_header; then
>         swagger_targets+=" scylla_swagger_gen_$name"
>       fi
>     fi
>   done
>   cmake                                         \
>     --build build                               \
>      --target seastar_http_request_parser       \
>      --target idl-sources                       \
>      --target $swagger_targets
>   shell: sh -e {0}
>   env:
>     BUILD_TYPE: RelWithDebInfo
>     BUILD_DIR: build
>     CLEANER_OUTPUT_PATH: build/clang-include-cleaner.log
>     CLEANER_DIRS: test/unit exceptions alternator api auth cdc compaction
> ninja: error: unknown target 'scylla_swagger_gen_cql_server_test'
> Error: Process completed with exit code 1.
> ```
> 
> @yaronkaikov

@tchaikov Can you have a look on this failure, should we ignore it?

tchaikov commented on configure.py:
```diff
@@ -1234,6 +1234,8 @@ def find_ninja():
        Json2Code('api/api-doc/authorization_cache.json'),
        'api/raft.cc',
        Json2Code('api/api-doc/raft.json'),
+       Json2Code('api/api-doc/cql_server_test.json'),
```
@Jadw1 hi Michal, could you please

1.  update `api/CMakeLists.cmake` accordingly, namely, to add  `api-doc/cql_server_test.json` right after `api-doc/config.json`.
2. and add `cql_server_test.cc` right after `config.cc`

? thanks.

tchaikov commented on api/cql_server_test.cc:
```diff
@@ -0,0 +1,63 @@
+        , _timeout(timeout) {}
+    
+    std::string to_json() const override {
+        return fmt::format("{{\"role_name\":\"{}\", \"workload_type\":\"{}\", \"timeout\":\"{}\"}}", _role_name, _workload_type, _timeout);
```
i'd suggest use the raw string literal here, like:
```c++
return fmt::format(R"({{"role_name":"{}", "workload_type":"{}", "timeout":"{}"}})",
                   _role_name, _workload_type, _timeout);
```

tchaikov commented:
> > `iwyu / Analyze #includes in source files (pull_request) ` failed but the error seems unrelated to the PR code
> > ```
> >   swagger_targets=''
> >   for f in api/api-doc/*.json; do
> >     if test "${f#*.}" = json; then
> >       name=$(basename "$f" .json)
> >       if test $name != swagger20_header; then
> >         swagger_targets+=" scylla_swagger_gen_$name"
> >       fi
> >     fi
> >   done
> >   cmake                                         \
> >     --build build                               \
> >      --target seastar_http_request_parser       \
> >      --target idl-sources                       \
> >      --target $swagger_targets
> >   shell: sh -e {0}
> >   env:
> >     BUILD_TYPE: RelWithDebInfo
> >     BUILD_DIR: build
> >     CLEANER_OUTPUT_PATH: build/clang-include-cleaner.log
> >     CLEANER_DIRS: test/unit exceptions alternator api auth cdc compaction
> > ninja: error: unknown target 'scylla_swagger_gen_cql_server_test'
> > Error: Process completed with exit code 1.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @yaronkaikov
> 
> @tchaikov Can you have a look on this failure, should we ignore it?

@yaronkaikov thanks for the ping. we'd better not. otherwise i have to create another PR to fix the cmake build.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10741/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10741/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: boost/service_level_controller_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10741/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2902/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10924/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10741/)
#### Build Details:
- Duration: 5 hr 9 min
- Builder: spider8.cloudius-systems.com

Jadw1 commented:
Changelog:
- resolved merge conflict
- changed `assert()` to `SCYLLA_ASSERT()`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10776/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10776/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: boost/service_level_controller_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10776/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2923/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10945/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10776/)
#### Build Details:
- Duration: 5 hr 16 min
- Builder: spider2.cloudius-systems.com

Jadw1 commented:
@piodul ping

piodul commented on api/cql_server_test.cc:
```diff
@@ -0,0 +1,63 @@
+
+}
+
+#endif
```
Ditto

piodul commented on api/cql_server_test.hh:
```diff
@@ -0,0 +1,29 @@
+
+}
+
+#endif
```
Ditto

piodul commented on api/cql_server_test.cc:
```diff
@@ -0,0 +1,63 @@
+        , _timeout(timeout) {}
+    
+    std::string to_json() const override {
+        return fmt::format(R"({{"role_name":"{}", "workload_type":"{}", "timeout":"{}"}})", _role_name, _workload_type, _timeout);
```
Hmm, I'm worried about escaping the strings here.

I see that `json::json_base` exists, which looks like a much safer interface for objects that need to be serialized. Can we use it here?

piodul commented on transport/server.cc:
```diff
@@ -2037,4 +2037,20 @@ future<> cql_server::update_connections_service_level_params() {
+        auto& client_state = cql_conn.get_client_state();
+        auto& user = client_state.user();
+        auto role_name = user 
+                ? (user->name ? *(user->name) : "ANNONYMOUS") 
```
Typo: `ANNONYMOUS` -> `ANONYMOUS`.

It was pointed out by the spellchecker...

piodul commented on test/auth_cluster/test_raft_service_levels.py:
```diff
@@ -170,3 +171,139 @@ async def test_service_levels_work_during_recovery(manager: ManagerClient):
+    elif mode == "debug":
+        return "3m"
+    else:
+        "shouldn't be reached"
```
If this path can't be reached, let's put an assertion here.

piodul commented on test/auth_cluster/test_raft_service_levels.py:
```diff
@@ -170,3 +171,139 @@ async def test_service_levels_work_during_recovery(manager: ManagerClient):
+
+@pytest.mark.asyncio
+async def test_connections_parameters_auto_update(manager: ManagerClient, mode):
+    if mode == 'release':
+        # Skip the test in release mode because
+        # `/cql_server_test/connections_params` endpoint is not present there.
+        return
```
How about using the `@skip_mode` fixture?

piodul commented on test/auth_cluster/test_raft_service_levels.py:
```diff
@@ -170,3 +171,139 @@ async def test_service_levels_work_during_recovery(manager: ManagerClient):
+    cql = manager.get_cql()
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+
+    logging.info("Creating user roles and theirs connections")
```
Nit: `their connections`

piodul commented on test/auth_cluster/test_raft_service_levels.py:
```diff
@@ -170,3 +171,139 @@ async def test_service_levels_work_during_recovery(manager: ManagerClient):
+        connection = cluster.connect()
+        sessions.append((cluster, connection))
+    
+    return sessions
```
I suggest separating clusters and sessions into two separate lists and return them. It's not clear that a list called "sessions" contains pairs of cluster and session objects.

piodul replied:
Perhaps returning sessions is not necessary at all if you don't use them? Maybe we can just return cluster objects.

Jadw1 replied:
Sessions aren't used directly but they hold users' connections.

piodul reviewed (changes requested):
There are three typos detected by the spellchecker, please fix them,

Jadw1 commented:
Changelog:
Addressed comments and fixed typos.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10823/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10823/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: auth_cluster/test_auth_v2_migration
:small_blue_diamond: auth_cluster/test_raft_service_levels
:small_blue_diamond: boost/service_level_controller_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10823/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2963/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10823/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10985/)
#### Build Details:
- Duration: 4 hr 28 min
- Builder: spider1.cloudius-systems.com

piodul commented:
> > @Jadw1 merging this PR will fix #12923, right?
> > Yes

I updated the cover letter so that the PR fixes this issue.

piodul commented:
Queued

Annamikhlin commented:
@Jadw1 / @avikivity - after merge from OSS to enterprise 2 auth dtests started to fail constantly and prevent next-enterprise to pass.

- https://jenkins.scylladb.com/job/scylla-enterprise/job/next/3085/testReport/auth_test/TestAuth/Build___x86___dtest___test_transitional_auth_betweenness_from_default/

- https://jenkins.scylladb.com/job/scylla-enterprise/job/next/3085/testReport/auth_test/TestAuth/Build___x86___dtest___test_transitional_auth_betweenness_from_pwdauth/

failed with the same error:
```
self = <auth_test.TestAuth object at 0x7f8f4ae9b710>

    @pytest.mark.next_gating
    def test_transitional_auth_betweenness_from_default(self):
        """
        Start cluster with default Auth, test user permission during rolling upgrade of enable Transitional Auth.
        """
        logger.info("STEP: start cluster with default AllowAllAuthenticator/AllowAllAuthorizer")
        self.prepare(nodes=2, enable_auth=False)
        nodes = self.cluster.nodelist()
    
        logger.info("STEP: update config and restart node1 to enable Transitional Auth")
        nodes[0].stop(wait_other_notice=True, gently=True)
        config = {"authenticator": "com.scylladb.auth.TransitionalAuthenticator", "authorizer": "com.scylladb.auth.TransitionalAuthorizer"}
        nodes[0].set_configuration_options(values=config)
        nodes[0].start(wait_for_binary_proto=True)
        timeout = 30 if self.cluster.scylla_mode != "debug" else 180
        wait_for_any_log(self.cluster.nodelist(), "Created default superuser authentication record", timeout)
    
        session = self.get_session(node_idx=0, user="cassandra", password="cassandra")
        session.execute("CREATE USER normal WITH PASSWORD '123456' NOSUPERUSER")
    
        logger.info("STEP: (on node1) verify normal user has permission to list users")
        session = self.get_session(node_idx=0, user="normal", password="123456")
        session.execute("LIST USERS")
        logger.info("STEP: (on node1) verify user will login as anonymous if authentication fails")
>       session = self.get_session(node_idx=0, user="normal", password="wrongpwd")

auth_test.py:1866: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
auth_test.py:2072: in get_session
    conn = self.patient_exclusive_cql_connection(node, user=user, password=password)
dtest_setup.py:567: in patient_exclusive_cql_connection
    return retry_till_success(
dtest_setup.py:189: in retry_till_success
    return fun(*args, **kwargs)
dtest_setup.py:427: in exclusive_cql_connection
    return self._create_session(node, keyspace, user, password, compression, protocol_version, port=port, ssl_opts=ssl_opts, load_balancing_policy=wlrr, **kwargs)
dtest_setup.py:496: in _create_session
    session = cluster.connect(wait_for_all_pools=True)
cassandra/cluster.py:1761: in cassandra.cluster.Cluster.connect
    ???
cassandra/cluster.py:1787: in cassandra.cluster.Cluster.connect
    ???
cassandra/cluster.py:1774: in cassandra.cluster.Cluster.connect
    ???
cassandra/cluster.py:3603: in cassandra.cluster.ControlConnection.connect
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.cluster.NoHostAvailable: ('Unable to connect to any servers', {'127.0.7.1:9042': ConnectionShutdown('Connection to 127.0.7.1:9042 was closed')})

cassandra/cluster.py:3664: NoHostAvailable
```

nyh commented:
It turns out that this PR solves (at least, in theory - I don't think anyone tested it) an important Alternator performance problem - https://github.com/scylladb/scylladb/issues/18738 - encountered by a real user. We should have at least considered a backport - I see backport/none without explanation.

nyh commented:
> It turns out that this PR solves (at least, in theory - I don't think anyone tested it) an important Alternator performance problem - #18738 - encountered by a real user. We should have at least considered a backport - I see backport/none without explanation.

Actually, it seems we would only need to backport a piece of this patch (the caching piece). I'm not sure how feasable this is :-(
